src/compiler/parser.rs

```diff
@@ -22,25 +25,58 @@ enum SpeculationKind {
     Reparse,
 }
 
-fn visit_node<TNodeCallback: FnMut(Rc<Node>)>(mut cb_node: TNodeCallback, node: Rc<Node>) {
+fn visit_node<TNodeCallback: FnMut(Option<Rc<Node>>)>(
+    mut cb_node: &mut TNodeCallback,
+    node: Option<Rc<Node>>,
+) {
     cb_node(node)
 }
 
-pub fn for_each_child<TNodeCallback: FnMut(Rc<Node>), TNodesCallback: FnMut(&NodeArray)>(
+fn visit_nodes<TNodeCallback: FnMut(Option<Rc<Node>>), TNodesCallback: FnMut(&NodeArray)>(
+    mut cb_node: TNodeCallback,
+    mut cb_nodes: TNodesCallback,
+    nodes: &NodeArray,
+) {
+    if true {
+        if true {
+            return cb_nodes(nodes);
+        }
+    }
+    unimplemented!()
+}
+
+pub fn for_each_child<TNodeCallback: FnMut(Option<Rc<Node>>), TNodesCallback: FnMut(&NodeArray)>(
     node: Rc<Node>,
-    cb_node: TNodeCallback,
+    mut cb_node: TNodeCallback,
     cb_nodes: TNodesCallback,
 ) {
     if node.kind() <= SyntaxKind::LastToken {
         return;
     }
     match &*node {
+        Node::VariableDeclaration(variable_declaration) => {
+            visit_node(&mut cb_node, Some(variable_declaration.name()));
```

Comment:
The compiler was complaining that this call of `visit_node()` takes ownership of the `cb_node` closure argument (and then I try and call `visit_node()` with it again right after this), which I guess makes sense, so I modified `visit_node()` to take a reference to a closure instead of a closure